packages/@lwc/engine-core/src/framework/vnodes.ts

```diff
@@ -34,7 +34,12 @@ export type VNode =
     | VFragment
     | VScopedSlotFragment;
 
-export type VNodes = Readonly<Array<VNode | null>>;
+export type VNodes = ReadonlyArray<VNode | null>;
+/**
+ * Mutable version of {@link VNodes}. It should only be used inside functions to build an array;
+ * it should never be used as a parameter or return type.
```

Comment:
I don't suppose there's anyway we can enforce this using TypeScript?

The whole point of making these read-only is because, due to the static object optimization (where e.g. `{ data: { foo: 'bar' } }` gets hoisted out of the `tmpl` function into a global constant), you can't really rely on a VNodes array being mutable.

Reply:
Mutable arrays are a superset of readonly arrays, so TypeScript will happily accept a readonly array. It's [technically possible](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggTnAhiAsgV2I4BLA9gOwGcoBeKAUQA8BjAGzQBMIAeAawhFwDMo19X8uAO74A2gF0ANFHacecCIgYFaIXv0EiJAPgBQoSFAByEAG4Q46TACNa0MgAVEcHIlrMAShGq44DZvBIqBhYeETS+GYW2noG0ADKwHDY1MBeSiogzAAqUBCUwBD4DMQKGfiq6gLCYuLapFDpyhVZ2fUAZMZRliG2ELq6XHypYVCCVoh9DnaIhCy5+YXFpYrNlXzVWnUAFM5wAFxQicmpTZk52gCUUADeAL4DPkTAUGUMAPItDaIAjNIATNIAMziKCzKBPQjAXTjXp2aaKObbN6fVTXAD06LGuDyCF8UEAvBuAWR3dJCXgBbOH2KC-AHA8Qw3ATKYzJGUmx2DFYixwfHE3RAA) to have TypeScript enforce it, but that kind of check feels more linter-level to me. (Even this PR is probably overkill for the complexity vs value added.)